fw/mmap_buffer.c

```diff
@@ -210,29 +184,57 @@ tfw_mmap_buffer_create(const char *filename, unsigned int size)
 	if (filename && strlen(filename) >= TFW_MMAP_BUFFER_MAX_NAME_LEN - 1)
 		return NULL;
 
-	holder = kmalloc(sizeof(TfwMmapBufferHolder) +
-			 sizeof(struct page *) * num_online_cpus(),
+	if (holders_cnt + 1 > MAX_HOLDERS)
+		return NULL;
+
+	holder = kzalloc(sizeof(TfwMmapBufferHolder) +
+			 sizeof(TfwMMapBufferMem) * num_online_cpus(),
 			 GFP_KERNEL);
 	if (!holder)
 		return NULL;
 
 	order = get_order(size);
 
 	holder->dev_major = -1;
+	holder->size = size;
 	holder->buf = (TfwMmapBuffer **)alloc_percpu_gfp(sizeof(TfwMmapBuffer *),
-													 GFP_KERNEL);
+							 GFP_KERNEL);
 	atomic_set(&holder->dev_is_opened, 0);
 	atomic_set(&holder->is_freeing, 0);
 
+	page_cnt = size / PAGE_SIZE;
+	/*
+	 * Allocate pages for double mapping and a page for buffer control structure.
+	 */
+	page_ptr = kmalloc(page_cnt * sizeof(struct page *) * 2 + 1, GFP_KERNEL);
+	if (!page_ptr)
+		return NULL;
```

Comment:
It seems previously allocated `holder` leaked here

Reply:
I noticed and fixed it when fixed holder->buf handling.